packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts

```diff
@@ -92,44 +93,70 @@ export class ModularChangeFamily
         return { fieldKind, changesets: normalizedChanges };
     }
 
-    compose(changes: FieldChangeMap[]): FieldChangeMap {
-        if (changes.length === 1) {
-            return changes[0];
-        }
-
+    compose(changes: TaggedChange<FieldChangeMap>[]): FieldChangeMap {
         const fieldChanges = new Map<FieldKey, FieldChange[]>();
         for (const change of changes) {
-            for (const [key, fieldChange] of change) {
-                getOrAddEmptyToMap(fieldChanges, key).push(fieldChange);
+            for (const [key, fieldChange] of change.change) {
+                const fieldChangeToCompose =
+                    fieldChange.revision !== undefined || change.revision === undefined
+                        ? fieldChange
+                        : {
+                              ...fieldChange,
+                              revision: change.revision,
+                          };
+
+                getOrAddEmptyToMap(fieldChanges, key).push(fieldChangeToCompose);
             }
         }
 
         const composedFields: FieldChangeMap = new Map();
         for (const [field, changesForField] of fieldChanges) {
-            const { fieldKind, changesets } = this.normalizeFieldChanges(changesForField);
-            const composedField = fieldKind.changeHandler.rebaser.compose(
-                changesets,
-                this.childComposer,
-            );
+            let composedField: FieldChange;
+            if (changesForField.length === 1) {
+                composedField = changesForField[0];
```

Comment:
Does this not break the composeChild contract? (And if not, shouldn't we be cloning this?)

Reply:
I don't think this breaks the contract. `changesForField[0]` was created in the loop above (line 100), and should be annotated with the revision it came from (unless it's already a composite, in which case its descendents are already annotated approrpriately). Since there are no other changes in the field, we can use `changesForField[0]` in the output, as it recursively specifies which revision all descdents are from.

If we want to avoid including any objects from the input in our output, then we would need to clone the change.

Reply:
I just realized: about the `composeChild` contract, this code is exempt because it's part of the modular change family (as opposed to a field kind implementation of compose).
And yeah, I momentarily forgot we're just shooting for "no mutation of the input" as opposed to "no objects included from the input in the output".